src/api/Consumer.js

```diff
@@ -0,0 +1,125 @@
+const _ = require('highland')
+const { merge } = require('tinyfunk')
+const { once } = require('ramda')
+const { tapP } = require('@articulate/funky')
+
+// Consumer :: Object -> { start, stop }
+const Consumer = db => opts => {
+  const {
+    batchSize = 1000,
+    category,
+    handlers = {},
+    groupMember,
+    groupSize,
+    init,
+    name,
+    positionUpdateInterval = 100,
+    tickInterval = 100
+  } = opts
+
+  let count = 0
+  let position = 0
+  const suffix = groupMember ? `-${groupMember}:${groupSize}` : ''
+  const streamName = `${category}+position-${name}${suffix}`
+  let up = false
+
+  const pollOpts = { batchSize, category, groupMember, groupSize }
+
+  const debug =
+    require('../lib/debug').extend(`consumer-${name}${suffix}`)
+
+  const assignPosition = msg => {
+    position = msg ? msg.data.position : 0
+  }
+
+  const cleanup = once(signal => {
+    debug(`received ${signal}, stopping`)
+    up = false
+  })
+
+  const handleMessage = msg =>
+    typeof handlers[msg.type] === 'function'
+      ? _(tapP(handlers[msg.type])(msg))
+      : _.of(msg)
+
+  const loadPosition = () =>
+    db.getLastStreamMessage(streamName).then(assignPosition)
+
+  const logMessage = msg =>
+    debug('dispatching message: %o', msg)
+
+  const poll = () => {
+    const stream = db.getCategoryMessages(merge(pollOpts, { position }))
+
+    stream.observe().each(logMessage)
+
+    stream.flatMap(handleMessage)
+      .flatMap(updatePosition)
+      .stopOnError(stop)
+      .done(tick)
+  }
+
+  const start = async () => {
+    if (up) {
+      debug('already started')
+      return
+    }
+
+    debug('starting on category: %o', category)
+    debug('handling types: %o', Object.keys(handlers))
+
+    if (typeof init === 'function') {
+      await init()
+      debug('initialized')
+    }
+
+    await loadPosition()
+    debug('position loaded: %o', position)
+
+    up = true
+    poll()
+  }
+
+  const stop = err => {
+    if (up) {
+      debug(`stopping${err ? ' on error' : ''}`)
+      up = false
+    } else {
+      debug('already stopped')
+    }
+
+    if (err) throw err
```

Comment:
My goal was to throw an `UncaughtException`.  We want to kill the process, otherwise the consumer will stop processing, but the process (and host) will stay running.

Reply:
But it's an out-of-band exception, right? It's in a stream, so `try/catch` won't find it. There's no way to catch & handle this error, unless i'm missing something?

Reply:
That's correct.  When we hit an unhandled error processing a message, we want to stop and die, not catch.  The time to catch and remediate that error is inside a handler, if appropriate.  If it's bubbled all the way back to here, then it's unexpected, and we should stop and get a human involved.  Otherwise, we'll have skipped a message in the stream, and all future written messages and read models will not reflect the true state of the message store.

There is a little bit more writeup on this in the [Eventide Error Handling docs](http://docs.eventide-project.org/user-guide/consumers.html#error-handling), and the `Error Handling` section of the API docs in this PR.

Reply:
After reading both sets of docs again, i think i'm going to double-down. An uncaught exception should _never_ be intended. Though i understand this is a fatal, non-streamable error. 

So maybe a koa-style error emitter? If there is an `'error'` listener, we emit the error. If not, we throw.

Reply:
An uncaught exception is - by definition - not intended.  Otherwise we would have caught it.  But we should make a distinction between an *uncaught* error, and an *unreported* error.  If we throw, node will [emit an `uncaughtException` event](https://devdocs.io/node/process#process_event_uncaughtexception), which [`bugsnag` will automatically report](https://docs.bugsnag.com/platforms/javascript/#reporting-unhandled-errors).